### PR TITLE
[ONNX] Export embedding_bag 

### PR DESCRIPTION
**Pull request by neginraoof**
Enable export of embedding_bag op to ONNX

### CODE REVIEW

File: torch/onnx/symbolic_opset10.py [hunk 1]
```diff
@@ -178,3 +179,67 @@ def flip(g, input, dims):
...
+                  offsets,
+                  scale_grad_by_freq,
+                  mode,
+                  sparse,
+                  per_sample_weights,
+                  include_last_offset):
+    from torch.onnx.symbolic_opset9 import size, div
+
+    if not sym_help._is_none(per_sample_weights):
+        return sym_help._onnx_unsupported('embedding_bag  with per_sample_weights')
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py]:
```suggestion
        return sym_help._onnx_unsupported('embedding_bag with per_sample_weights')
```
nit space

File: torch/onnx/symbolic_opset10.py [hunk 2]
```diff
@@ -178,3 +179,67 @@ def flip(g, input, dims):
...
+                else:
+                    embeddings = g.op("ReduceMax", embeddings, axes_i=[0], keepdims_i=0)
+
+                embeddings = g.op("Unsqueeze", embeddings, axes_i=[0])
+                list_.append(embeddings)
+
+        output = g.op("Concat", *list_, axis_i=0)
+        return output, None, None, None
+    else:
+        return sym_help._onnx_unsupported('embedding_bag')
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py:249]: add more info to this message :)

File: torch/onnx/symbolic_opset10.py [hunk 3]
```diff
@@ -178,3 +179,67 @@ def flip(g, input, dims):
...
+
+    if offsets.type().sizes() is not None:
+        if indices.node().kind() != 'prim::Param':
+            embeddings = g.op("Gather", embedding_matrix, indices)
+            dim_0 = size(g, offsets, g.op("Constant", value_t=torch.LongTensor([0])))
+            dim_1 = div(g, size(g, indices, g.op("Constant", value_t=torch.LongTensor([0]))),
+                        size(g, offsets, g.op("Constant", value_t=torch.LongTensor([0]))))
+            dim_2 = size(g, embedding_matrix, g.op("Constant", value_t=torch.LongTensor([1])))
+
+            shape = [dim_0, dim_1, dim_2]
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py:206]: i think `dim_1` can be replaced with constant -1

File: torch/onnx/symbolic_opset10.py [hunk 4]
```diff
@@ -178,3 +179,67 @@ def flip(g, input, dims):
...
+            shape = g.op("Concat", *shape, axis_i=0)
+
+            embeddings = g.op("Reshape", embeddings, shape)
+            if mode == 0:
+                embeddings = g.op("ReduceSum", embeddings, axes_i=[1], keepdims_i=0)
+            elif mode == 1:
+                embeddings = g.op("ReduceMean", embeddings, axes_i=[1], keepdims_i=0)
+            else:
+                embeddings = g.op("ReduceMax", embeddings, axes_i=[1], keepdims_i=0)
+            return embeddings, None, None, None
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py:220]: do you know why there are four outputs for the `embedding_bag` schema?

File: torch/onnx/symbolic_opset10.py [hunk 5]
```diff
@@ -178,3 +179,67 @@ def flip(g, input, dims):
...
+                  indices,
+                  offsets,
+                  scale_grad_by_freq,
+                  mode,
+                  sparse,
+                  per_sample_weights,
+                  include_last_offset):
+    from torch.onnx.symbolic_opset9 import size, div
+
+    if not sym_help._is_none(per_sample_weights):
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py]: Add similar check for `include_last_offset`.

`scale_grad_by_freq` is for training, maybe add check as well if exporting in training mode?

File: torch/onnx/symbolic_opset10.py [hunk 6]
```diff
@@ -178,3 +179,67 @@ def flip(g, input, dims):
...
+        return sym_help._onnx_unsupported('embedding_bag  with per_sample_weights')
+
+    if offsets.type().sizes() is not None:
+        if indices.node().kind() != 'prim::Param':
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py]: I assume this is the 2d input case. But why can we tell from this condition? `if indices.node().kind() != 'prim::Param':`

Also worth noting here why while the doc mentions `offsets` should be None, but still have value here.

> If input is 2D of shape (B, N),
> 
> it will be treated as B bags (sequences) each of fixed length N, and this will return B values aggregated in a way depending on the mode. **offsets is ignored and required to be None in this case.**
>

Comment (neginraoof) [reply torch/onnx/symbolic_opset10.py]: I actually updated this to check for onnx::Reshape type. All inputs (1D and 2D) are reshaped to 1D in functional.py. And for 2D inputs offsets is set to:
torch.arange(0, indices.numel(), indices.size(1))

File: torch/onnx/symbolic_opset10.py (hunk 4 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset10.py]: I added a comment explaining that the three last outputs are unused in torch.nn modules.

**Review by BowenBao — APPROVED** [review]
LGTM

**Comment by neginraoof** [comment]
This PR is approved within our team. This is ready for review form @houseroad.

**Review by bzinodev — APPROVED** [review]

**Review by houseroad — COMMENTED** [review]
Could you rebase the PR to resolve the conflict?
